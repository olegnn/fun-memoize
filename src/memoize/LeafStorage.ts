import { AbsentValue, NO_VALUE } from "../value";
import { ChildPath, Storage, StorageParams } from "../base/Storage";
import { CacheStrategy } from "../base/CacheStrategy";
import { Result } from "../strategy";
import { SizedIterable } from "../iterators";
import { Destroyable } from "../utils";

/**
 * Leaf storage callbacks.
 */
export interface LeafStorageParams<K, V> extends StorageParams<K, V> {
  /** Callback to be called on the leaf creation */
  onCreateLeaf?: (leafStorage: K) => void;
  /** Callback to be called on the leaf removal */
  onRemoveLeaf?: (leafStorage: K) => void;
}

/**
 * Stores leaf key -> value pairs.
 */
export class LeafStorage<K, V> extends Storage<K, V> implements Destroyable {
  declare params: LeafStorageParams<K, V>;
  storage: Storage<K, V>;
  strategy: CacheStrategy<K>;
  dropStorageValue: (key: K) => boolean;

  constructor(
    storage: Storage<K, V>,
    strategy: CacheStrategy<K>,
    params: LeafStorageParams<K, V>,
    rootPath?: Iterable<ChildPath<K>>
  ) {
    super(params, rootPath);
    this.storage = storage;
    this.strategy = strategy;
    this.params = params;
    this.dropStorageValue = this.storage.drop.bind(this.storage);
  }

  /**
   * Returns amount of keys (references) stored in a map.
   *
   */
  len(): number {
    return this.storage.len();
  }

  /**
   * Drops an item corresponding to the supplied key.
   * @param key
   *
   */
  drop(key: K): V | AbsentValue {
    const isWeak = this.storage.isWeak(key);
    if (!isWeak) {
      const dropped = this.strategy.drop(key);

      if (dropped && this.params.onRemoveLeaf != null)
        this.params.onRemoveLeaf(key);
    }
    const removedFromStorage = this.storage.drop(key);
    if (isWeak) {
      if (removedFromStorage !== NO_VALUE && this.params.onRemoveLeaf != null)
        this.params.onRemoveLeaf(key);
    }

    return removedFromStorage;
  }

  /**
   * Retrieves an item corresponding to the supplied key.
   * @param key
   *
   */
  get(key: K): V | AbsentValue {
    const isWeak = this.storage.isWeak(key);
    const extracted = this.storage.get(key);
    if (!isWeak && extracted !== NO_VALUE) {
      this.handleResult(this.strategy.read(key));
    }

    return extracted;
  }

  /**
   * Associates supplied item with the key.
   * @param key
   * @param value
   *
   */
  set(key: K, value: V): void {
    const isWeak = this.storage.isWeak(key);
    if (!isWeak) this.handleResult(this.strategy.write(key));

    this.storage.set(key, value);
  }

  /**
   * Removes an item from the beginning of the queue.
   *
   */
  remove(): K | AbsentValue {
    const key = this.strategy.remove();
    if (key !== NO_VALUE) {
      this.storage.drop(key as K);

      if (this.params.onRemoveLeaf != null) this.params.onRemoveLeaf(key as K);
    }

    return key;
  }

  /**
   * Calls a `destroy` implementations that will unlink given storage from all entities
   * referencing it for both storage and cache strategy.
   *
   */
  destroy(): void {
    super.destroy();
    this.strategy.destroy();
  }

  /**
   * Removes all items from the storage and cache strategy.
   *
   */
  clear(): void {
    this.storage.clear();
    this.strategy.clear();
  }

  /**
   * Returns an iterator over the entries.
   *
   */
  entries(): SizedIterable<{ key: K; value: V }> {
    return this.storage.entries();
  }

  /**
   * Executes callbacks for each added/removed item.
   * @param result
   */
  private handleResult(result: Result<K>) {
    result.forEachRemoved(this.dropStorageValue);

    if (this.params.onCreateLeaf != null) {
      result.forEachAdded(this.params.onCreateLeaf);
    }

    if (this.params.onRemoveLeaf != null) {
      result.forEachRemoved(this.params.onRemoveLeaf);
    }
  }
}
