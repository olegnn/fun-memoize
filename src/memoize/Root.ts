import { equals, AbsentValue, NO_VALUE } from "../value";
import { ChildPath, Storage } from "../base/Storage";
import { append, double, once } from "../iterators";
import { EMPTY_ARRAY } from "../utils";
import { LeafStorage } from "./LeafStorage";
import { StorageContext, NestedStorage } from "./StorageContext";

/**
 * Contains either a value or a pointer.
 */
class ResultOrPointer<R, P> {
  result: R | AbsentValue;
  ptr: P | undefined;

  private constructor() {
    this.result = NO_VALUE;
    this.ptr = void 0;
  }

  /**
   * Produces a result.
   * @param result
   */
  static fromResult<R>(result: R | AbsentValue): ResultOrPointer<R, unknown> {
    const that = new ResultOrPointer();
    that.result = result;

    return that;
  }

  /**
   * Produces a pointer.
   * @param ptr
   */
  static fromPointer<P>(ptr: P): ResultOrPointer<unknown, P> {
    const that = new ResultOrPointer();
    that.ptr = ptr;

    return that as ResultOrPointer<unknown, P>;
  }
}

/**
 * Weakly or strongly linked value.
 */
class LinkedValue<V> {
  /** Underlying value. */
  value: V;
  /** Is value a weekly linked? */
  weak: boolean;

  constructor(value: V, weak: boolean) {
    this.value = value;
    this.weak = weak;
  }
}

/**
 * Stores last path to value including storage nodes.
 * Used to reduce the number of access operations in case the prefix has some elements in common.
 */
class Last<K, V> {
  /**
   * Last storage path.
   */
  storages: Array<LinkedValue<NestedStorage<K, V>>>;
  /**
   * Last key path.
   */
  path: Array<K>;
  /**
   * Last extracted value.
   */
  value: V | AbsentValue;

  constructor(length: number, root: NestedStorage<K, V>) {
    this.storages = Array(length);
    this.storages[0] = new LinkedValue(root, false);
    this.path = EMPTY_ARRAY;
    this.value = NO_VALUE;
  }

  /**
   * Associates storage with the provided index.
   * @param idx
   * @param storage
   */
  setStorage(idx: number, storage: LinkedValue<NestedStorage<K, V>>) {
    this.storages[idx] = storage;
  }

  /**
   * Retrieves storage using supplied index.
   * @param idx
   */
  storage(idx: number): LinkedValue<NestedStorage<K, V>> {
    return this.storages[idx];
  }

  /**
   * Returns a value in case the path is the same as the previous one.
   * Otherwise returns an index of the greatest common node for provided and stored paths.
   * @param path
   */
  get(path: K[]): ResultOrPointer<V, number> {
    const length = this.path.length;
    if (length === path.length) {
      let idx = -1;
      for (let i = 0; i < length && equals(this.path[i], path[i]); i++) {
        idx = i;
      }

      if (idx === length - 1) {
        return ResultOrPointer.fromResult(this.value) as ResultOrPointer<
          V,
          number
        >;
      }

      return ResultOrPointer.fromPointer(idx === -1 ? 0 : idx);
    } else {
      return ResultOrPointer.fromPointer(0);
    }
  }

  /**
   * Updates path and stored value.
   * @param path
   * @param value
   */
  update(path: K[], value: V) {
    this.path = path;
    this.value = value;
  }
}

export class Root<K, V> {
  /**
   * Cache depth.
   */
  length: number;
  /**
   * Root node.
   */
  root: NestedStorage<K, V>;
  /**
   * Storage context used to create new storages.
   */
  ctx: StorageContext<K, V>;

  /**
   * Root storage strategy child path to pass to newly created nodes.
   */
  rootPath: ChildPath<AbsentValue>;
  /**
   * Root leaf storage strategy child path to pass to newly created leaf nodes.
   */
  leafPath: ChildPath<AbsentValue>;

  /**
   * Last path, storage path, and value are stored to reduce the amount of access operations for keys having common prefixes.
   */
  last: Last<K, V>;

  constructor(length: number, ctx: StorageContext<K, V>) {
    this.length = length;
    this.ctx = ctx;
    this.root = (
      length > 1 ? ctx.createStorage() : ctx.createLeafStorage()
    ) as NestedStorage<K, V>;
    this.last = new Last(length, this.root);

    this.rootPath = new ChildPath(this.ctx.rootStorageStrategy, NO_VALUE);
    this.leafPath = new ChildPath(this.ctx.rootLeafStrategy, NO_VALUE);
  }

  /**
   * Retrieves cached value if already calculated, otherwise calls the supplied function with provided arguments and
   * saves the result.
   * @param path
   * @param calculate
   */
  getOrInsertWith(path: K[], calculate: (args: K[]) => V): V {
    const length = this.length;
    if (path.length !== length) throw new Error("Invalid path length");

    let { result, ptr } = this.last.get(path);
    if (result !== NO_VALUE) {
      this.readCache();

      return result as V;
    }

    ({ result, ptr } = this.extractPath(path, ptr));
    if (result !== NO_VALUE) {
      this.last.update(path, result as V);
      this.readCache();

      return result as V;
    }

    result = calculate.apply(null, path);
    const value = this.setPath(path, result as V, ptr);
    this.last.update(path, value);
    this.writeCache();

    return value;
  }

  private extractPath(path: K[], from: number = 0): ResultOrPointer<V, number> {
    const length = this.length;
    const node = this.last.storage(from);
    let idx = from,
      lessThanPath = true,
      res: V | AbsentValue | NestedStorage<K, V> = node.value,
      isLastWeak = node.weak;

    do {
      this.last.setStorage(
        idx,
        new LinkedValue(res as NestedStorage<K, V>, isLastWeak)
      );

      isLastWeak = (res as Storage<K, V>).isWeak(path[idx]);
      res = (res as Storage<K, V>).get(path[idx]);

      lessThanPath = ++idx < length;
    } while (lessThanPath && res !== NO_VALUE);

    if (lessThanPath || res === NO_VALUE) {
      return ResultOrPointer.fromPointer(Math.max(0, idx - 1));
    }

    return ResultOrPointer.fromResult(res) as ResultOrPointer<V, number>;
  }

  private setPath(path: Array<K>, value: V, from: number = 0) {
    const length = this.length;
    let idx = from,
      cache: NestedStorage<K, V> = this.last.storage(from).value;

    while (idx < length - 1) {
      const current = path[idx];
      let next = cache.get(current);
      const isWeak = cache.isWeak(current);

      if (next === NO_VALUE) {
        const rootPath = isWeak
          ? once(this.rootPath)
          : double(this.rootPath, new ChildPath(cache, current));
        next =
          idx < length - 2
            ? this.ctx.createStorage(rootPath)
            : this.ctx.createLeafStorage(append(rootPath, this.leafPath));

        cache.set(current, next as unknown as V);
      }

      cache = next as NestedStorage<K, V>;
      this.last.setStorage(++idx, new LinkedValue(cache, isWeak));
    }

    cache.set(path[length - 1], value);

    return value;
  }

  private readCache() {
    for (let i = this.length; i-- > 0; ) {
      const node = this.last.storage(i);
      if (!node.weak) this.ctx.rootStorageStrategy.read(node.value);
    }

    const leafNode = this.last.storage(this.length - 1);
    if (!leafNode.weak) {
      this.ctx.rootLeafStrategy.read(leafNode.value as LeafStorage<K, V>);
    }
  }

  private writeCache() {
    for (let i = this.length; i-- > 0; ) {
      const node = this.last.storage(i);
      if (!node.weak) this.ctx.rootStorageStrategy.write(node.value);
    }

    const leafNode = this.last.storage(this.length - 1);
    if (!leafNode.weak) {
      this.ctx.rootLeafStrategy.write(leafNode.value as LeafStorage<K, V>);
    }
  }
}
