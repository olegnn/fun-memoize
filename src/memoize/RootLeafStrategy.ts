import { CacheStrategy } from "../base/CacheStrategy";
import { once } from "../iterators";
import { Result } from "../strategy/types";
import { AbsentValue, NO_VALUE } from "../value";
import { LeafStorage } from "./LeafStorage";

/**
 * Cache strategy for leaf values.
 */
export class RootLeafStrategy<K, V> extends CacheStrategy<LeafStorage<K, V>> {
  _strategy: CacheStrategy<LeafStorage<K, V>>;
  _children: number;

  constructor(
    leafCapacity: number,
    strategy: CacheStrategy<LeafStorage<K, V>>
  ) {
    super(leafCapacity);
    this._strategy = strategy;
    this._children = 0;
  }

  /**
   * Records a read access of the supplied value.
   */
  read(value: LeafStorage<K, V>): Result<LeafStorage<K, V>> {
    return super
      .read(value)
      .chain(this.handleInnerStrategyResult(this._strategy.read(value)));
  }

  /**
   * Records a write access of the supplied value.
   */
  write(value: LeafStorage<K, V>): Result<LeafStorage<K, V>> {
    this._children++;

    return super
      .write(value)
      .chain(this.handleInnerStrategyResult(this._strategy.write(value)));
  }

  /**
   * Reserves place for an item.
   * @param value
   */
  protected reservePlace(_: LeafStorage<K, V>): Result<LeafStorage<K, V>> {
    if (this.isFull()) {
      const removedItem = this.remove();

      if (removedItem !== NO_VALUE) {
        return Result.removed(once(removedItem as LeafStorage<K, V>));
      }
    }

    return Result.empty();
  }

  /**
   * Removes supplied value from the cache.
   */
  drop(value: LeafStorage<K, V>): boolean {
    const dropped = this._strategy.drop(value);
    if (dropped) {
      this._children -= (value as LeafStorage<K, V>).len();
    }

    return dropped;
  }

  /**
   * Returns underlying amount of tracked cache entries.
   */
  len(): number {
    return this._children;
  }

  /**
   * Returns `true` if supplied node exists in cache.
   */
  has(node: LeafStorage<K, V>) {
    return this._strategy.has(node);
  }

  /**
   * Peeks an item from the beginning of the queue.
   */
  peek() {
    return this._strategy.peek();
  }

  /**
   * Removes an item from the beginning of the queue.
   *
   */
  remove(): LeafStorage<K, V> | AbsentValue {
    const maybeStorage = this.peek();

    if (maybeStorage !== NO_VALUE) {
      const storage = maybeStorage as LeafStorage<K, V>;
      const removed = storage.remove();
      if (removed !== NO_VALUE) {
        this._children--;
      }

      if (storage.isEmpty()) {
        this.drop(storage);

        return storage;
      }
    }

    return NO_VALUE;
  }

  /**
   * Handles a result produced by calling one of the inner strategy methods.
   */
  private handleInnerStrategyResult(result: Result<LeafStorage<K, V>>) {
    for (const removed of result.removed) {
      this._children -= removed.len();
    }

    return result;
  }
}
