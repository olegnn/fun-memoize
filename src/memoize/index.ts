import mimic from "mimic-fn";
import { Params, StorageContext } from "./StorageContext";
import { Root } from "./Root";
import { equals, AbsentValue, NO_VALUE } from "../value";
import { EMPTY_OBJECT } from "../utils";

/** Params interface extended with optional length and checkLast flag */
export interface ParamsWithLength<K, V> extends Params<K, V> {
  /** Overrides function length */
  length?: number;
  /** Check last arguments or not (default to `true`) */
  checkLast?: boolean;
}

const { slice } = Array.prototype;

/**
 * Memoizes provided function returning wrapped version of it.
 * Result function will return value without calling the supplied function if it's present in the cache for the supplied arguments according to `Same-value-zero` algorithm.
 * If no value is found, the underlying function will be called with provided arguments.
 * @param func
 * @param params
 */
export default function memoize<K, V>(
  func: (...args: K[]) => V,
  {
    length = func.length,
    checkLast = true,
    ...params
  }: ParamsWithLength<K, V> = EMPTY_OBJECT as ParamsWithLength<K, V>
): typeof func {
  const recomputate = function () {
    ++(resultFunction as any).recomputations;
    return func.apply(func, arguments);
  };

  let resultFunction: typeof func;
  if (length === 0) {
    let value = NO_VALUE;

    resultFunction = function cachedFunction() {
      if (value !== NO_VALUE) {
        return value as V;
      } else {
        return (value = recomputate.apply(this, arguments));
      }
    };
  } else {
    const ctx = new StorageContext(params);
    const root = new Root(length, ctx);

    resultFunction = function cachedFunction() {
      const argsLength = arguments.length;
      let output: V;

      if (argsLength === length) {
        output = root.getOrInsertWith(arguments as unknown as K[], recomputate);
        // If we received a greater amount of arguments, slice it
      } else if (argsLength > length) {
        output = root.getOrInsertWith(
          slice.call(arguments, 0, length) as K[],
          recomputate
        );
        // Otherwise don't use cache
      } else {
        output = recomputate.apply(func, arguments);
      }

      return output;
    };
  }

  if (checkLast) {
    const fn = resultFunction;
    let lastCache: V | AbsentValue = NO_VALUE;
    let lastArgs: IArguments = [] as unknown as IArguments;

    resultFunction = function cachedFunction(): V {
      let i = arguments.length;
      if (i === lastArgs.length)
        if (i === 1) {
          if (equals(arguments[0], lastArgs[0]) && lastCache !== NO_VALUE)
            return lastCache as V;
        } else {
          while (i-- && equals(arguments[i], lastArgs[i]));
          if (i === -1 && lastCache !== NO_VALUE) return lastCache as V;
        }
      lastArgs = arguments;

      return (lastCache = fn.apply(null, arguments));
    };
  }

  try {
    void mimic(resultFunction, func);
  } finally {
    (resultFunction as any).recomputations = 0;
    return resultFunction;
  }
}
