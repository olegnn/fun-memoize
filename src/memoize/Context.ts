import {
  CacheStrategy,
  CacheStrategyClass,
  withDestroyable,
} from "../base/CacheStrategy";
import { EMPTY_ITERABLE } from "../iterables";
import { LeafStorage, LeafStorageParams } from "./LeafStorage";
import { Storage, StorageClass, StorageParams } from "../base/Storage";
import {
  SafeMapStorage,
  UnifiedStorage,
  UnifiedStorageParams,
} from "../storage";
import { RootLeafStrategy } from "./RootLeafStrategy";
import { LRU, Noop } from "../strategy";
import { DEFAULT_MAX_ENTRIES_COUNT } from "../constants";
import { noop, ParentPath } from "../utils";

/**
 * Either leaf storage or nested storage containing either nested storages or leaf storages.
 */
export type NestedStorage<K, V> =
  | LeafStorage<K, V>
  | Storage<K, NestedStorage<K, V> | LeafStorage<K, V> | V>;

/**
 * Params for the storage context.
 */
export interface Params<K, V>
  extends UnifiedStorageParams<K, V>,
    LeafStorageParams<K, V>,
    StorageParams<K, V> {
  /**
   * Total limit for the storages (cache nodes).
   */
  totalStoragesLimit?: number;
  /**
   * Total limit for the leaves (cache entries). Default is 10000.
   */
  totalLeavesLimit?: number;
  /**
   * Limit of the leaves per a single leaf storage.
   */
  leavesPerStorageLimit?: number;
  /**
   * Total limit of the leaf storages.
   */
  totalLeafStoragesLimit?: number;
  /**
   * Either strategy class or different strategy classes for leaves and storage nodes.
   */
  strategy?:
    | StrategyConfig<K, V>
    | CacheStrategyClass<K | LeafStorage<K, V> | NestedStorage<K, V>>;
}

/**
 * Config for the leaf and storage cache strategies.
 */
type StrategyConfig<K, V> = {
  leafStrategyClass: CacheStrategyClass<K | LeafStorage<K, V>>;
  storageStrategyClass: CacheStrategyClass<NestedStorage<K, V>>;
};

/**
 * Picks a `Noop` strategy if the `limit` is `Infinity`.
 * @param strategy
 * @param limit
 */
const pickNoopIfNoLimit = <V>(
  strategy: CacheStrategyClass<V>,
  limit: number
): CacheStrategyClass<V> => (limit !== Infinity ? strategy : Noop);

/**
 * Checks whether the provided limit is a valid number.
 * @param limit
 */
const limitIsValid = (limit: number) =>
  (Number.isSafeInteger(limit) && limit > 0) || limit === Infinity;

/**
 * Ensures that the provided limit is a valid number.
 * @param name
 * @param value
 * @param min
 */
const ensureLimitIsValid = (name: string, value: number, min: number) => {
  if (!limitIsValid(value)) {
    throw new Error(`\`${name}\` is invalid: \`${value}\``);
  }
  if (value < min) {
    throw new Error(
      `\`${name}\`'s value = \`${value}\` is too small, expected \`${min}\` at least`
    );
  }
};

/**
 * Returns `true` if supplied class extends `CacheStrategy`.
 * @param strategy
 */
const extendsCacheStrategy = (strategy: unknown) =>
  strategy &&
  (
    strategy as {
      prototype: any;
    }
  ).prototype instanceof CacheStrategy;

/**
 * Class describing a cache strategy to be used for the leaf nodes.
 */
type LeafCacheStrategyClass<K, V> = CacheStrategyClass<LeafStorage<K, V>>;
/**
 * Class describing a cache strategy to be used for the storage nodes.
 */
type StorageCacheStrategyClass<K, V> = CacheStrategyClass<NestedStorage<K, V>>;

/**
 * Params with cache depth.
 */
type ParamsWithDepth<K, V> = Params<K, V> & {
  /**
   * Cache depth.
   */
  depth: number;
};

/**
 * Root storage context.
 */
export class Context<K, V> {
  /**
   * Strategy containing all storage nodes.
   */
  rootStoragesStrategy: CacheStrategy<NestedStorage<K, V>>;
  /**
   * Strategy containing all leaf storages.
   */
  rootLeafStoragesStrategy: CacheStrategy<LeafStorage<K, V>>;
  /**
   * Params for all storages.
   */
  params: LeafStorageParams<K, V> & UnifiedStorageParams<K, V>;
  /**
   * Limit of leaf per a single leaf storage.
   */
  leavesPerStorageLimit: number;

  /**
   * Storage class to be used for cached values.
   */
  storageClass: StorageClass<K, V>;
  /**
   * Strategy class used for leaf values.
   */
  leafStrategyClass: CacheStrategyClass<K>;

  constructor({
    depth,
    strategy = LRU,
    totalLeavesLimit = DEFAULT_MAX_ENTRIES_COUNT,
    totalStoragesLimit = Infinity,
    totalLeafStoragesLimit = Infinity,
    leavesPerStorageLimit = Infinity,
    useWeakStorage = false,
    useObjectStorage = false,
    onCreateStorage = noop,
    onRemoveStorage = noop,
    onCreateLeaf = noop,
    onRemoveLeaf = noop,
  }: ParamsWithDepth<K, V>) {
    if (!Number.isSafeInteger(depth) || depth <= 0) {
      throw new TypeError(
        `Invalid depth, an expected natural number which is a safe integer`
      );
    }
    let leafStrategyClass: LeafCacheStrategyClass<K, V>,
      storageStrategyClass: StorageCacheStrategyClass<K, V>;
    if (extendsCacheStrategy(strategy)) {
      leafStrategyClass = strategy as LeafCacheStrategyClass<K, V>;
      storageStrategyClass = strategy as StorageCacheStrategyClass<K, V>;
    } else {
      const config = strategy as StrategyConfig<K, V>;

      leafStrategyClass = config.leafStrategyClass as LeafCacheStrategyClass<
        K,
        V
      >;
      storageStrategyClass =
        config.storageStrategyClass as StorageCacheStrategyClass<K, V>;
    }
    if (!extendsCacheStrategy(leafStrategyClass)) {
      throw new Error("Invalid `leafStrategyClass`");
    }
    if (!extendsCacheStrategy(storageStrategyClass)) {
      throw new Error("Invalid `storageStrategyClass`");
    }
    ensureLimitIsValid("totalLeavesLimit", totalLeavesLimit, 1);
    ensureLimitIsValid("totalStoragesLimit", totalStoragesLimit, depth);
    ensureLimitIsValid("totalLeafStoragesLimit", totalLeafStoragesLimit, 1);
    ensureLimitIsValid("leavesPerStorageLimit", leavesPerStorageLimit, 1);

    storageStrategyClass = withDestroyable(
      pickNoopIfNoLimit(storageStrategyClass, totalStoragesLimit)
    );
    leafStrategyClass =
      totalLeavesLimit === totalLeafStoragesLimit &&
      totalLeavesLimit === leavesPerStorageLimit
        ? pickNoopIfNoLimit(leafStrategyClass, totalLeavesLimit)
        : leafStrategyClass;

    this.rootStoragesStrategy = new storageStrategyClass(
      totalStoragesLimit - 1
    );
    this.rootLeafStoragesStrategy = new RootLeafStrategy(
      totalLeavesLimit,
      new (withDestroyable(leafStrategyClass))(totalLeafStoragesLimit)
    );
    this.storageClass = (
      useWeakStorage || useObjectStorage ? UnifiedStorage : SafeMapStorage
    ) as StorageClass<K, V>;
    this.leafStrategyClass = leafStrategyClass as CacheStrategyClass<K>;
    this.leavesPerStorageLimit = leavesPerStorageLimit;
    this.params = {
      useObjectStorage,
      useWeakStorage,
      onCreateStorage,
      onRemoveStorage,
      onCreateLeaf,
      onRemoveLeaf,
    };
  }

  /**
   * Instantiates new storage with supplied parents
   * @param root
   */
  createStorage(
    parentPaths: Iterable<ParentPath<K | NestedStorage<K, V>>> = EMPTY_ITERABLE
  ): NestedStorage<K, V> {
    return new this.storageClass(this.params, parentPaths);
  }

  /**
   * Instantiates new leaf storage with supplied parents
   * @param root
   */
  createLeafStorage(
    parentPaths: Iterable<ParentPath<K | LeafStorage<K, V>>> = EMPTY_ITERABLE
  ): LeafStorage<K, V> {
    return new LeafStorage(
      this.createStorage() as Storage<K, V>,
      new this.leafStrategyClass(this.leavesPerStorageLimit),
      this.params,
      parentPaths
    );
  }
}
