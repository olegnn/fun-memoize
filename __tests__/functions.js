import mimic from 'mimic-fn';
import memoize from '../src';

const { describe, test, expect } = global;

const createFunctionForTests = (func) => {
  const calledWith = new Set();
  const newFunc = (...args) => calledWith.has(String(args)) ? void 0: (calledWith.add(String(args)), func(...args));
  mimic(newFunc, func);
  return newFunc;
};

describe(
  'Basic function memoize tests', () => {
    test(
      'Memoize simple function', () => {
        const func = a => a * 125;

        expect(memoize(func)(2)).toBe(250);
      },
    );

    test(
      'Memoize unary function', () => {
        const f = n => n * 15;
        const func = createFunctionForTests(f);

        const memoized = memoize(func);
        const testNumbers = [3, 4, 5, 6, 7, 8, 9];
        for (let i = 1e2; i--;)
          expect(testNumbers.map(memoized)).toEqual(
            testNumbers.map(f),
          );
      },
    );

    test(
      'Memoize 2ary function', () => {
        const f = (a, b) => a + b;
        const func = createFunctionForTests(f);

        const memoized = memoize(func);
        const testArgs = [
          [1, 2],
          [6, 9],
          [1, 2],
          [10, 12],
          [20, 22],
        ];
        for (let i = 1e2; i--;)
          expect(testArgs.map(arr => memoized(...arr))).toEqual(
            testArgs.map(arr => f(...arr)),
          );
      },
    );

    test(
      'Memoize 5ary function', () => {
        const f = (a, b, c, d, e) => Math.max(...[a, b, c, d, e].map(v => v && v.length || 0));
        const func = createFunctionForTests(f);

        const memoized = memoize(func);
        const testArgs = [
          [1, 2, 'hello world', 18, 22],
          [6, 'evil', 'hello', 'typeof', 0],
          [[], { length: 2e2 }, [], {}, null],
          [() => {}, { length: 1 }, [1111], {}, null],
        ];
        for (let i = 1e2; i--;)
          expect(testArgs.map(arr => memoized(...arr))).toEqual(
            testArgs.map(arr => f(...arr)),
          );
      },
    );
  },
);
