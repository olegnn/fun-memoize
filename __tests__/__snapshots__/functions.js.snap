// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic function memoize tests different limits 1`] = `"\`totalLeavesLimit\` is invalid: \`-1\`"`;

exports[`Basic function memoize tests different limits 2`] = `"\`totalLeavesLimit\` is invalid: \`-Infinity\`"`;

exports[`Basic function memoize tests different limits 3`] = `"\`totalLeavesLimit\` is invalid: \`0\`"`;

exports[`Basic function memoize tests different limits 4`] = `"\`totalLeavesLimit\` is invalid: \`123456789123456780\`"`;

exports[`Basic function memoize tests different limits 5`] = `"\`totalLeavesLimit\` is invalid: \`1.231231\`"`;

exports[`Basic function memoize tests different limits 6`] = `"\`totalLeavesLimit\` is invalid: \`NaN\`"`;

exports[`Basic function memoize tests different limits 7`] = `"\`totalStoragesLimit\` is invalid: \`-1\`"`;

exports[`Basic function memoize tests different limits 8`] = `"\`totalStoragesLimit\` is invalid: \`-Infinity\`"`;

exports[`Basic function memoize tests different limits 9`] = `"\`totalStoragesLimit\` is invalid: \`0\`"`;

exports[`Basic function memoize tests different limits 10`] = `"\`totalStoragesLimit\` is invalid: \`123456789123456780\`"`;

exports[`Basic function memoize tests different limits 11`] = `"\`totalStoragesLimit\` is invalid: \`1.231231\`"`;

exports[`Basic function memoize tests different limits 12`] = `"\`totalStoragesLimit\` is invalid: \`NaN\`"`;

exports[`Basic function memoize tests different limits 13`] = `"\`leavesPerStorageLimit\` is invalid: \`-1\`"`;

exports[`Basic function memoize tests different limits 14`] = `"\`leavesPerStorageLimit\` is invalid: \`-Infinity\`"`;

exports[`Basic function memoize tests different limits 15`] = `"\`leavesPerStorageLimit\` is invalid: \`0\`"`;

exports[`Basic function memoize tests different limits 16`] = `"\`leavesPerStorageLimit\` is invalid: \`123456789123456780\`"`;

exports[`Basic function memoize tests different limits 17`] = `"\`leavesPerStorageLimit\` is invalid: \`1.231231\`"`;

exports[`Basic function memoize tests different limits 18`] = `"\`leavesPerStorageLimit\` is invalid: \`NaN\`"`;

exports[`Basic function memoize tests different limits 19`] = `"\`totalLeafStoragesLimit\` is invalid: \`-1\`"`;

exports[`Basic function memoize tests different limits 20`] = `"\`totalLeafStoragesLimit\` is invalid: \`-Infinity\`"`;

exports[`Basic function memoize tests different limits 21`] = `"\`totalLeafStoragesLimit\` is invalid: \`0\`"`;

exports[`Basic function memoize tests different limits 22`] = `"\`totalLeafStoragesLimit\` is invalid: \`123456789123456780\`"`;

exports[`Basic function memoize tests different limits 23`] = `"\`totalLeafStoragesLimit\` is invalid: \`1.231231\`"`;

exports[`Basic function memoize tests different limits 24`] = `"\`totalLeafStoragesLimit\` is invalid: \`NaN\`"`;
